### ISSUE TITLE
Rename ruleset to Grammar?

### ISSUE THREAD

**Issue by tawheeler** [sisl/ExprRules.jl#5]
It is a grammar, I feel we might want to call it a grammar.

One caveat is that a grammar has a set of starting symbols.

**Comment by rcnlee** [comment]
I second this recommendation.

Typically :start is defined in the grammar rather than in the function calls.  I think that would better align it with the existing literature.

**Comment by mykelk** [comment]
I'm okay with this.

**Comment by tawheeler** [comment]
Is it better to have a single start symbol or a set of start symbols?

**Comment by mykelk** [comment]
Without loss of generality, we can just assume one.

**Comment by mykelk** [comment]
I discussed in more detail with @tawheeler and arrived at the following... Let's just rename `@ruleset` to `@grammar`. We don't need to specify the start symbol in the grammar; only in the function where it is used. We should allow for an array of start symbols. @rcnlee let us know if you think differently.

**Comment by rcnlee** [comment]
That's fine to declare the start symbol at point-of-use.  

Wouldn't it be cleaner to just support one start symbol?  Otherwise, every function that uses a start symbol will need to be defined twice and there will be code duplication.  For example,
``function StatsBase.sample(root::RuleNode, typ::Symbol, ruleset::RuleSet)``
``function StatsBase.sample(root::RuleNode, typs::Vector{Symbol}, ruleset::RuleSet)``

We shouldn't need to support that as a special case.  The user can define it in the grammar if they want a concatenation of start symbols.

**Comment by mykelk** [comment]
I think you can just do something like this:
`StatsBase.sample(root::RuleNode, typ::Symbol, ruleset::RuleSet) = sample(root, [typ], ruleset)`
and then put the implementation in the `Vector` version. Right?

**Comment by rcnlee** [comment]
What is the expected behavior of functions with an array of start symbols?  Does it return one tree or an array of trees?
What is the use case?

**Comment by mykelk** [comment]
If you're sampling, just one tree. @tawheeler might be able to elaborate on the use case.

**Comment by tawheeler** [comment]
If we need a set of start symbols we can do it the other way around:
```
StatsBase.sample(root::RuleNode, types::AbstractVector{Symbol}, ruleset::RuleSet) = sample(root, sample(types), ruleset)
```
to avoid allocation.

Maybe I was wrong - most places only seem to use a single start symbol. I personally don't have much use for a set of start symbols.

**Comment by mykelk** [comment]
We can just go with a single start symbol if there isn't a need for a vector. I think we'd just need to make a couple small changes to the book chapter.

**Comment by rcnlee** [comment]
Renamed ruleset to grammar throughout.  Single start symbol at point-of-use is already existing.

Closed by 069d4cac19dd08c51a0c9d3ba8de3ad5af861afb
